src/dune_cache/trimmer.ml

```diff
@@ -0,0 +1,115 @@
+open Stdune
+open Dune_cache_storage
+
+module Trimming_result = struct
+  type t = { trimmed_bytes : int64 }
+
+  let empty = { trimmed_bytes = 0L }
+
+  (* CR-someday amokhov: Right now Dune doesn't support large (>1Gb) files on
+     32-bit platforms due to the pervasive use of [int] for representing
+     individual file sizes. It's not fundamentally difficult to switch to
+     [int64], so we should do it if it becomes a real issue. *)
+  let add t ~(bytes : int) =
+    { trimmed_bytes = Int64.add t.trimmed_bytes (Int64.of_int bytes) }
+end
+
+let trim_broken_metadata_entries ~trimmed_so_far =
+  List.fold_left Version.Metadata.all ~init:trimmed_so_far
+    ~f:(fun trimmed_so_far version ->
+      let metadata_entries = Layout.Versioned.list_metadata_entries version in
+      let file_path =
+        Layout.Versioned.file_path (Version.Metadata.file_version version)
+      in
+      List.fold_left metadata_entries ~init:trimmed_so_far
+        ~f:(fun trimmed_so_far path ->
+          let should_be_removed =
+            match Digest.from_hex (Path.basename path) with
```

Comment:
Suggestion: this could be a function in `Layout`. Or `list_metadata_entries` could return something structured, like a list of:

```
{ rule_or_action_digest : Digest.t
; metadata_file : Path.t
}
```

Reply:
Sounds good, I'll change this.

Reply:
Done.